### PR TITLE
EAGLE parallel draft with auto regression; kv cache in EAGLE training

### PR DESCRIPTION
**Pull request by yeyu-nvidia**
## What does this PR do?

**Type of change:** ? new feature

**Overview:** 
This PR adds 2 features:
1) Integrate parallel draft with auto regression in EAGLE training and inference. Now when parallel_draft_step > 1, multiple tokens are generated in each step in training/inference. All parallel draft tokens will be used as context for the next ttt_step training.
2) Use kv cache in EAGLE training. The attention mask is no longer a seq-len**2 square mask, but a q-len*k-len rectangle mask. This reduces memory consumption and avoid redundant computation. 

## Usage
No API change.

## Testing
Offline tested with Qwen3-30B and Llama3.2-1B. Passed nmm-sandbox regression test.

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Summary by CodeRabbit

* **New Features**
  * Inference context can be passed through speculative decoding and top-level generation flows; public entry points accept an optional inference context.
  * Multi-draft generation now supports per-draft/per-step progression with per-step KV/cache updates, cumulative loss, and per-draft accuracy tracking.

* **Refactor**
  * Replaced fixed multi-step assembly with loop-based per-step construction for inputs, attention masks, and sequence-parallel handling to simplify draft propagation.

* **Chores**
  * Cross-rank consistency checks now warn by default; strict failure is opt-in.

### CODE REVIEW

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 1]
```diff
@@ -889,178 +780,58 @@ def _get_eagle_module_inputs(
...
                 )
-            else:
-                # [TODO] (yeyu): there will be problem here with MLA
-                eagle_inputs["rotary_pos_emb"] = None
-        else:
-            eagle_inputs["input_ids"] = torch.cat(
-                (padded_input_ids, padded_input_ids, padded_input_ids, padded_input_ids),
-                dim=-1,
             )
+        )
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/megatron_eagle.py]: Previous draft tokens are already considered in the kv cache

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 2]
```diff
@@ -1260,20 +1046,28 @@ def forward(
...
+                    eagle_inputs_0,
+                    output_weight,
+                    inference_params=inference_params,
+                    packed_seq_params=packed_seq_params,
+                    inference_context=eagle_inference_context,
+                    **(extra_block_kwargs or {}),
+                )
+
+                eagle_logits_0.append(eagle_logits_)
+            eagle_logits_0 = torch.cat(eagle_logits_0, dim=0)
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/megatron_eagle.py]: This is expected for parallel draft

File: modelopt/torch/speculative/plugins/megatron_eagle.py [hunk 3]
```diff
@@ -1230,6 +1033,12 @@ def forward(
             output_weight = self.shared_embedding_or_output_weight()
         logits_sbh, _ = self.output_layer(hidden_states, weight=output_weight)
 
+        # EAGLE kv cache
+        eagle_inference_context = StaticInferenceContext(
+            input_ids.shape[0],
+            input_ids.shape[1] * (self.eagle_config.parallel_draft_step + ttt_steps - 1),
+        )
```
[the remark this answers is not in the document]
Comment (yeyu-nvidia) [reply modelopt/torch/speculative/plugins/megatron_eagle.py:1013]: No, the original value is correct

**Comment by yeyu-nvidia** [comment]
Closing as stale. This branch dates to 2025 and has long since diverged from main (conflicting), and the EAGLE/parallel-draft work it explores has been superseded by the current speculative-decoding pipeline in main. Reopen if any of it is still wanted — nothing here is lost, the branch remains.
